op-node/rollup/driver/state.go

```diff
@@ -440,6 +440,10 @@ func (s *state) eventLoop() {
 			} else if err != nil && errors.Is(err, derive.ErrCritical) {
 				s.log.Error("Derivation process critical error", "err", err)
 				return
+			} else if err != nil && errors.Is(err, derive.NotEnoughData) {
+				stepAttempts = 0 // don't do a backoff for this error
+				reqStep()
+				continue
```

Comment:
This is quite subtle, but preserves one of the best things from the old design I think was most undervalued in redesign work.

While the pipeline is deriving, a stage can choose to stop and return this error, handing back control to the caller (driver). This keeps the derivation pipeline nice and synchronous for Cannon usage, while handling cases where there's a slow background fetching thing or something requiring repeated smaller updates (like here with the channels). And thus we can put a short and strict timeout on the pipeline work, while keeping the whole thing snappy and simple.

The alternative would be to make the pipeline blocking and essentially halt until we pull either a batch or an EOF, but then it would hold the Engine control captive unless we decouple that further (and deal with queuing and async stuff there without breaking Cannon compatibility), which I think isn't worth it (but maybe with a good design can be done eventually at a later point, since it's not a breaking protocol change).

cc @norswap this is related to thing we started to discuss.

Reply:
This error was mainly introduced to simplify some of the code (so it doesn't need to do a loop inside a function). Honestly, I'm really skeptical about the benefit of yielding back to the caller - it tends to introduce a lot of extra internal state to be able to resume and I'm not sure what the benefit of this is. Having a block pipeline is really simple and easy to understand. In addition, because the fetcher is just an interface, it's really easy to sub out for cannon.

Reply:
The pipeline blocking means we can't interact with the engine from the driver without knowing if we interfere with the engine state or if things change while we use the engine, handing back control to the caller to not lock the engine is important IMHO. Agreed on the blocking being easier, but I don't believe we should block on something too long when it wraps a resource as important as the engine.